src/openstack_cloud/openstack_runner_manager.py

```diff
@@ -0,0 +1,688 @@
+            )
+            return None
+
+        metric = runner_metrics.extract(
+            metrics_storage_manager=metrics_storage, runners=instance.server_name
+        )
+        self._delete_runner(instance, remove_token)
+        return next(metric, None)
+
+    def cleanup(self, remove_token: str) -> Iterator[runner_metrics.RunnerMetrics]:
+        """Cleanup runner and resource on the cloud.
+
+        Args:
+            remove_token: The GitHub remove token.
+
+        Returns:
+            Any metrics retrieved from cleanup runners.
+        """
+        runners = self._get_runner_health()
+        healthy_runner_names = [runner.server_name for runner in runners.healthy]
+        metrics = runner_metrics.extract(
+            metrics_storage_manager=metrics_storage, runners=set(healthy_runner_names)
+        )
+        for runner in runners.unhealthy:
+            self._delete_runner(runner, remove_token)
+
+        self._openstack_cloud.cleanup()
+        return metrics
+
+    def _delete_runner(self, instance: OpenstackInstance, remove_token: str) -> None:
+        """Delete self-hosted runners by openstack instance.
+
+        Args:
+            instance: The OpenStack instance.
+            remove_token: The GitHub remove token.
+        """
+        try:
+            ssh_conn = self._openstack_cloud.get_ssh_connection(instance)
+            self._pull_runner_metrics(instance.server_name, ssh_conn)
+
+            try:
+                OpenstackRunnerManager._run_github_runner_removal_script(
+                    instance.server_name, ssh_conn, remove_token
+                )
+            except GithubRunnerRemoveError:
+                logger.warning(
+                    "Unable to run github runner removal script for %s",
+                    instance.server_name,
+                    stack_info=True,
+                )
+        except SshError:
+            logger.exception(
+                "Failed to get SSH connection while removing %s", instance.server_name
+            )
+            logger.warning(
+                "Skipping runner remove script for %s due to SSH issues", instance.server_name
+            )
+
+        try:
+            self._openstack_cloud.delete_instance(instance.instance_id)
+        except OpenstackError:
+            logger.exception(
+                "Unable to delete openstack instance for runner %s", instance.server_name
+            )
+
+    def _get_runner_health(self) -> RunnerHealth:
+        """Get runners by health state.
+
+        Returns:
+            Runners by health state.
+        """
+        runner_list = self._openstack_cloud.get_instances()
+
+        healthy, unhealthy = [], []
+        for runner in runner_list:
+            cloud_state = CloudRunnerState.from_openstack_server_status(runner.status)
+            if cloud_state in (
+                CloudRunnerState.DELETED,
+                CloudRunnerState.ERROR,
+                CloudRunnerState.STOPPED,
```

Comment:
Should UNKNOWN also be here?

Reply:
Fixed

Reply:
After thinking, UNKNOWN might not belong here. UNKNOWN might be returned if OpenStack does not have information on the server (e.g., unable to contact the OpenStack host for that server). Which does not mean the server is having issues.
Health check is performed for other states, and I think UNKNOWN should be a part of this.